server/sandbox_run_linux.go

```diff
@@ -37,6 +39,208 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/types"
 )
 
+// addToMappingsIfMissing ensures the specified id is mapped from the host.
+func addToMappingsIfMissing(ids []idtools.IDMap, id int64) []idtools.IDMap {
+	firstAvailable := int(0)
+	for _, r := range ids {
+		if int(id) >= r.HostID && int(id) < r.HostID+r.Size {
+			// Already present, nothing to do
+			return ids
+		}
+		if r.ContainerID+r.Size > firstAvailable {
+			firstAvailable = r.ContainerID + r.Size
+		}
+	}
+	newMapping := idtools.IDMap{
+		ContainerID: firstAvailable,
+		HostID:      int(id),
+		Size:        1,
+	}
+	return append(ids, newMapping)
+}
+
+func (s *Server) configureSandboxIDMappings(mode string, sc *pb.LinuxSandboxSecurityContext) (*storage.IDMappingOptions, error) {
+	if mode == "" {
+		// No mode specified but mappings set in the config file, let's use them.
+		if s.defaultIDMappings != nil {
+			uids := s.defaultIDMappings.UIDs()
+			gids := s.defaultIDMappings.GIDs()
+			return &storage.IDMappingOptions{UIDMap: uids, GIDMap: gids}, nil
+		}
+		return nil, nil
+	}
+
+	// expect a configuration like: private:uidmapping=0:1000:2000,2000:1000:2000;gidmapping=0:1000:4000,4000:1000:2000
+	parts := strings.SplitN(mode, ":", 2)
+
+	values := map[string]string{}
+	if len(parts) > 1 {
+		for _, r := range strings.Split(parts[1], ";") {
+			kv := strings.SplitN(r, "=", 2)
+			if len(kv) != 2 {
+				return nil, errors.Errorf("invalid argument: %q", r)
+			}
+			values[kv[0]] = kv[1]
+		}
+	}
+
+	switch parts[0] {
```

Comment:
Can we share this parsing between Podman and Cri-o?

Reply:
I am afraid we will end up with a different set of options for CRI-O.

I am still not sure how `auto` will look like, all the assumptions we had with Podman are not valid anymore.  We cannot really inspect the image since it is not known at user namespace creation time.

Reply:
Well we can just allocate the 65k and be done with it.  For now, if user does not specify it.

Kubernetes will keep better track of the containers and not leave them forever on disk.

Reply:
that is a good idea.  I think it should be either specified or default to something reasonable like 65k